File: apps/mobile/lib/core/logic/binder_names.dart
import '../models/binder.dart';

/// Result of create/rename name checks. [normalized] is trim-only (case kept).
class BinderNameResult {
  const BinderNameResult({
    required this.ok,
    required this.normalized,
    this.reason,
    this.role,
  });

  final bool ok;
  final String normalized;
  final String? reason;
  final BinderRole? role;
}

String normalizeBinderName(String raw) => raw.trim();

String _fold(String name) => normalizeBinderName(name).toLowerCase();

/// Validate a create or rename against live Binder names (trim + case-fold).
BinderNameResult validateBinderName({
  required String proposedName,
  required Iterable<Binder> binders,
  String? binderId,
}) {
  final normalized = normalizeBinderName(proposedName);
  if (normalized.isEmpty) {
    BinderRole? role;
    if (binderId != null) {
      for (final b in binders) {
        if (b.clientId == binderId) role = b.role;
      }
    }
    return BinderNameResult(
      ok: false,
      normalized: '',
      reason: 'empty',
      role: role,
    );
  }

  final folded = _fold(normalized);
  Binder? self;
  if (binderId != null) {
    for (final b in binders) {
      if (b.clientId == binderId) {
        self = b;
        break;
      }
    }
  }

  for (final b in binders) {
    if (!b.isLive) continue;
    if (binderId != null && b.clientId == binderId) continue;
    if (_fold(b.name) == folded) {
      return BinderNameResult(
        ok: false,
        normalized: normalized,
        reason: 'duplicate',
        role: self?.role,
      );
    }
  }

  return BinderNameResult(
    ok: true,
    normalized: normalized,
    role: self?.role ?? BinderRole.standard,
  );
}

File: apps/web/src/utils/binderNames.js
/**
 * Binder name uniqueness. Golden cases: packages/contracts/binder_names.json
 * Normalize: trim, then case-fold for comparison. Empty after trim is invalid.
 */

export function normalizeBinderName(raw) {
    return String(raw ?? '').trim();
}

function fold(name) {
    return normalizeBinderName(name).toLowerCase();
}

/**
 * @param {{ proposedName: string, binders: Array, binderId?: string|null }} args
 * @returns {{ ok: boolean, normalized: string, reason: string|null, role: string|null }}
 */
export function validateBinderName({ proposedName, binders, binderId = null }) {
    const normalized = normalizeBinderName(proposedName);
    const self = binderId
        ? (binders || []).find((b) => b.clientId === binderId)
        : null;
    const selfRole = self?.role || null;

    if (!normalized) {
        return { ok: false, normalized: '', reason: 'empty', role: selfRole };
    }

    const folded = fold(normalized);
    const collision = (binders || []).some((b) => {
        if (b.deletedAt) return false;
        if (binderId && b.clientId === binderId) return false;
        return fold(b.name) === folded;
    });
    if (collision) {
        return {
            ok: false,
            normalized,
            reason: 'duplicate',
            role: selfRole,
        };
    }
    return {
        ok: true,
        normalized,
        reason: null,
        role: selfRole || 'standard',
    };
}
